File: trees_and_graphs/lowest_common_ancestor.py
import treenode

def lowestCommonAncestorRecursive(root: treenode.TreeNode,
                                  p: treenode.TreeNode,
                                  q: treenode.TreeNode) -> treenode.TreeNode:
    if not root:
        return None
    
    if root == p or root == q:
        return root
    
    left = lowestCommonAncestorRecursive(root.left, p, q)
    right = lowestCommonAncestorRecursive(root.right, p, q)

    if left and right:
        return root
    
    if left:
        return left
    
    return right

File: trees_and_graphs/lowest_common_ancestor.java
public TreeNode lowestCommonAncestorRecursive(TreeNode root,
                                              TreeNode p,
                                              TreeNode q)
{
    if (root == null)
    {
        return null;
    }

    if (root == p || root == q)
    {
        return root;
    }

    TreeNode left = lowestCommonAncestorRecursive(root.left, p, q);
    TreeNode right = lowestCommonAncestorRecursive(root.right, p, q);

    if (left != null && right != null)
    {
        return root;
    }

    if (left != null)
    {
        return left;
    }

    return right;
}